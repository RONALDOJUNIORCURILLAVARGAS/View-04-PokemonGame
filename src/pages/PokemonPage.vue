<template>
  <div >
    <div class="head">
      <h1 class="life">Vidas : {{life}}</h1>
      <h1 class="points">Puntos : {{points}}</h1>
    </div>
    
    <h1 v-if="!pokemon">Espere por favor...</h1>
    <div v-else>
      <h1 >{{message_refresh}} </h1>
      <!-- TODO:img-->
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
      <!-- TODO:Opciones-->
    <PokemonOptions 
    :pokemons="pokemonArr"
    @selection-pokemon="checkAnswer($event)"
    />

    <template v-if="showAnswer" >
     <h2 class="fade-in">{{message}}</h2>
      
    </template>
    <template v-if="life===0">
      <button @click="refreshpage">
        Nuevo juego
      </button>
    </template>
    
    </div>
      
  </div>
</template>

<script>
//Number, required
import PokemonOptions from '../components/PokemonOptions.vue'
import PokemonPicture from '../components/PokemonPicture.vue'
import getPokemonOptions from '@/helpers/getPokemonOptions'
//console.log(getPokemonOptions())

export default {
name:'Pokemon Page',
components:{
    PokemonOptions,
    PokemonPicture
},
data() {
  return {
    pokemonArr:[],
    pokemon:null,
    showPokemon:false,
    showAnswer: false,
    message:'',
    message_refresh:'¿Quien es este pokémon?',
    life:3,
    points:0,
    counter_click:true
  }
},
methods: {
  async mixPokemonArray(){
    this.pokemonArr=await getPokemonOptions()
    //Agarramos un pokemos de los 4
    //de manera aleatoria
    const rndInt=Math.floor(Math.random()*4)
    this.pokemon=this.pokemonArr[rndInt]
    //console.log(this.pokemon)
  },
  checkAnswer(selectedid){
    if(this.counter_click){
          if(this.life>0){
            this.counter_click=false
            this.showPokemon=true
            this.showAnswer=true
            if(selectedid===this.pokemon.id){
              this.message=`Correcto, es ${this.pokemon.name}`
              this.points+=10
            }else{
              this.message=`Oops fallaste, era ${this.pokemon.name}`
              --this.life
            }
            //Despues de 3 segundos de ser ejecutado se recargara la pagina
            setTimeout(()=>{
              this.message_refresh='Recargando en  3...'
            },0)
            setTimeout(()=>{
              this.message_refresh='Recargando en  2...'
            },1000)
            setTimeout(()=>{
              this.message_refresh='Recargando en  1...'
            },2000)
            setTimeout(this.newGame, 3000)
        }
        else{
          this.showAnswer=true
          this.message='Ya no te quedan vidas, recarga la pagina mrd!'
        }
      }
      else{
        this.message='Habil te crees mongol, ya no puedes cambiar'
      
      }   

  },
  newGame(){
    this.showAnswer=false
    this.showPokemon=false
    this.pokemonArr=[]
    this.pokemon=null
    this.counter_click=true
    this.message_refresh='¿Quien es este pokémon?'
    this.mixPokemonArray()
    
  },
  refreshpage(){
    this.newGame()
    this.life=3
    this.points=0
    this.mixPokemonArray()
  }
},
//Es un metodo que se llama de una sola vez
mounted() {
  this.mixPokemonArray()
},

}
</script>
<style scoped>
.head{
  display: flex;
  justify-content: center;
}
.life{
  color: blue;
  width: 200px;
}
.points{
  right: 1rem;
  width: 200px;
}
</style>