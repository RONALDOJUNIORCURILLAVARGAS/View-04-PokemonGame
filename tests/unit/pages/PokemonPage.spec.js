import {shallowMount,mount} from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock';

describe('PokemonPage component', () => {
    let wrapper
    //let mixPokemonArraySpy
    beforeEach(()=>{
        //mixPokemonArraySpy=jest.spyOn(PokemonPage.methods,'mixPokemonArray')
        wrapper=shallowMount(PokemonPage)
    })
    test('debe de hace match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    })
    test('debe de llamar mixPokemonArray al montar', () => {
        const mixPokemonArraySpy=jest.spyOn(PokemonPage.methods,'mixPokemonArray')
        const wrapper=shallowMount(PokemonPage)
        expect(mixPokemonArraySpy).toHaveBeenCalled();
        //Esperamos que solo sea llamado una vez
        expect(mixPokemonArraySpy).toHaveBeenCalledTimes(1);
    })
    test('debe de hacer match con el snapshot cuando cargan los pokemon', () => {
        //El shallowMautn que monta de manera sutil
        //El mount monta todo
       
        const wrapper=mount(PokemonPage,{
            data() {
                return {
                  pokemonArr:pokemons,
                  pokemon:pokemons[0],
                  showPokemon:false,
                  showAnswer: false,
                  message:'',
                  message_refresh:'¿Quien es este pokémon?',
                  life:3,
                  points:0,
                  counter_click:true
                }
              }
        })
        expect(wrapper.html()).toMatchSnapshot();

    })
    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
        //mount
        const wrapper=shallowMount(PokemonPage,{
            data() {
                return {
                  pokemonArr:pokemons,
                  pokemon:pokemons[0],
                  showPokemon:false,
                  showAnswer: false,
                  message:'',
                  message_refresh:'¿Quien es este pokémon?',
                  life:3,
                  points:0,
                  counter_click:true
                }
              }
        })
        //PokemonPicture debe de existir
        const PokemonPicture=wrapper.find('pokemon-picture-stub')
        expect(PokemonPicture.exists()).toBeTruthy();

        //PokemonOption debe de existir
        const PokemonOptions = wrapper.find('pokemon-options-stub')
        expect(PokemonOptions.exists()).toBeTruthy()

        //PokemonPicture attribute pokemonId==5
        expect(PokemonPicture.attributes('pokemonid')).toBe('1');

        //PokemonOptions attribute pokemons toBe true
        const pokemones=PokemonOptions.attributes('pokemons')
        //expect(wrapper.find('pokemon-options-stub').attributes('pokemons')).toBeTruthy()
        expect(pokemones).toBeTruthy()
    })
    test('pruebas con CheckAnswer', async () => {
        const wrapper=shallowMount(PokemonPage,{
            data() {
                return {
                  pokemonArr:pokemons,
                  pokemon:pokemons[0],
                  showPokemon:false,
                  showAnswer: false,
                  message:'',
                  message_refresh:'¿Quien es este pokémon?',
                  life:3,
                  points:0,
                  counter_click:true
                }
              }
        })
        //Llamamos el metodo desde aqui
        console.log(wrapper.vm)
        //Como las propiedades son reactivas VUE tiene q esperar el cambio pór
        //eso se utilizara el await
        await wrapper.vm.checkAnswer(1)
        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correcto, es ${pokemons[0].name}`)
        await wrapper.vm.checkAnswer(5)
        expect(wrapper.vm.message).toBe(`Oops fallaste, era ${pokemons[0].name}`)
    })

    
    
    
    
});