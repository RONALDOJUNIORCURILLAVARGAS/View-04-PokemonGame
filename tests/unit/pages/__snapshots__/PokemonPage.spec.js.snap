// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component debe de hace match con el snapshot 1`] = `
<div>
  <div class="head">
    <h1 class="life">Vidas : 3</h1>
    <h1 class="points">Puntos : 0</h1>
  </div>
  <h1>Espere por favor...</h1>
</div>
`;

exports[`PokemonPage component debe de hacer match con el snapshot cuando cargan los pokemon 1`] = `
<div>
  <div class="head">
    <h1 class="life">Vidas : 3</h1>
    <h1 class="points">Puntos : 0</h1>
  </div>
  <div>
    <h1>¿Quien es este pokémon?</h1><!-- TODO:img-->
    <div>
      <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" class="hidden-pokemon" alt="pokemon"></div>
    </div><!-- TODO:Opciones-->
    <div>
      <div class="options-container">
        <ul>
          <li>pikachu</li>
          <li>venusaur</li>
          <li>charmeleon</li>
          <li>mew</li>
        </ul>
      </div>
    </div>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;
